# Fake News Detection 
Building a classifier differentiating fake and real news based on pre-defined labels

### Major Steps
1. Libraries
2. Loading and Cleaning datasets
3. Preparing dataset (BOW, TF-IDF)
4. SVM
5. Naive Bayes
6. Logistic Regression
7. PassiveAgressive Classifier
8. Summary

## 1. Import libraries

In [12]:
import pandas as pd
import numpy as np
import nltk
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

from sklearn.feature_extraction.text import CountVectorizer,  TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.metrics import accuracy_score

from sklearn.pipeline import Pipeline

In [13]:
import warnings
warnings.filterwarnings('ignore')

## 2. Load & Examine Datasets

In [14]:
df_train = pd.read_csv("/Users/taniaelachkar/Desktop/Natural Language Processing/Assignment 1/fake_or_real_news_training.csv")

df_test = pd.read_csv("/Users/taniaelachkar/Desktop/Natural Language Processing/Assignment 1/fake_or_real_news_test.csv")

In [15]:
df_train.head(20)

ID                                              title  \
0    8476                       You Can Smell Hillary’s Fear   
1   10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2    3608        Kerry to go to Paris in gesture of sympathy   
3   10142  Bernie supporters on Twitter erupt in anger ag...   
4     875   The Battle of New York: Why This Primary Matters   
5    6903                                        Tehran, USA   
6    7341  Girl Horrified At What She Watches Boyfriend D...   
7      95                  ‘Britain’s Schindler’ Dies at 106   
8    4869  Fact check: Trump and Clinton at the 'commande...   
9    2909  Iran reportedly makes new push for uranium con...   
10   1357  With all three Clintons in Iowa, a glimpse at ...   
11    988  Donald Trump’s Shockingly Weak Delegate Game S...   
12   7041  Strong Solar Storm, Tech Risks Today | S0 News...   
13   7623       10 Ways America Is Preparing for World War 3   
14   1571                   Trump takes on Cruz, but lightly   
15   4739                         How women lead differently   
16   7737  Shocking! Michele Obama & Hillary Caught Glamo...   
17   8716  Hillary Clinton in HUGE Trouble After America ...   
18   3304  What's in that Iran bill that Obama doesn't like?   
19   3078  The 1 chart that explains everything you need ...   

                                                 text label   X1   X2  
0   Daniel Greenfield, a Shillman Journalism Fello...  FAKE  NaN  NaN  
1   Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  NaN  NaN  
2   U.S. Secretary of State John F. Kerry said Mon...  REAL  NaN  NaN  
3   — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  NaN  NaN  
4   It's primary day in New York and front-runners...  REAL  NaN  NaN  
5     \nI’m not an immigrant, but my grandparents ...  FAKE  NaN  NaN  
6   Share This Baylee Luciani (left), Screenshot o...  FAKE  NaN  NaN  
7   A Czech stockbroker who saved more than 650 Je...  REAL  NaN  NaN  
8   Hillary Clinton and Donald Trump made some ina...  REAL  NaN  NaN  
9   Iranian negotiators reportedly have made a las...  REAL  NaN  NaN  
10  CEDAR RAPIDS, Iowa — “I had one of the most wo...  REAL  NaN  NaN  
11  Donald Trump’s organizational problems have go...  REAL  NaN  NaN  
12  Click Here To Learn More About Alexandra's Per...  FAKE  NaN  NaN  
13  October 31, 2016 at 4:52 am \nPretty factual e...  FAKE  NaN  NaN  
14  Killing Obama administration rules, dismantlin...  REAL  NaN  NaN  
15  As more women move into high offices, they oft...  REAL  NaN  NaN  
16  Shocking! Michele Obama & Hillary Caught Glamo...  FAKE  NaN  NaN  
17  0 \nHillary Clinton has barely just lost the p...  FAKE  NaN  NaN  
18  Washington (CNN) For months, the White House a...  REAL  NaN  NaN  
19  While paging through Pew's best data visualiza...  REAL  NaN  NaN

In [16]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
ID       3999 non-null int64
title    3999 non-null object
text     3999 non-null object
label    3999 non-null object
X1       33 non-null object
X2       2 non-null object
dtypes: int64(1), object(5)
memory usage: 187.5+ KB


### 2.1 Examine the X1 and X2 variables

In [17]:
df_train.X1.unique()

array([nan, 'REAL', 'FAKE',
       'PLANNED PARENTHOOD’S LOBBYING GETS AGGRESSIVE.\xa0Congress may have spent August away from Washington but Planned Parenthood’s campaign to convince lawmakers to protect the group’s funding followed them back to their home states. Power Post has more.\n\n“Lawmakers will raise the stakes when Congress returns next week by threatening to defund the group through the federal appropriations process. Planned Parenthood’s counter-offensive is widespread and varied and is unfolding inside and outside the Beltway. The group has been\xa0organizing rallies, flooding lawmakers’ town hall meetings, commissioning polls, shelling\xa0out six figures for television\xa0ads and\xa0hiring forensics experts to try to discredit undercover video footage that sparked the controversy. The success of these lobbying efforts will be tested when Congress returns and must move a short-term spending bill to keep the government open. Some conservatives in both chambers are pushing 

In [18]:
df_train.X2.unique()

array([nan, 'REAL', 'FAKE'], dtype=object)

In [19]:
#showing all the instances with X1
df_train[df_train.X1.notnull()]

ID                                              title  \
192     599                 Election Day: No Legal Pot In Ohio   
308   10194  Who rode it best? Jesse Jackson mounts up to f...   
382     356                     Black Hawk crashes off Florida   
660    2786     Afghanistan: 19 die in air attacks on hospital   
889    3622  Al Qaeda rep says group directed Paris magazin...   
911    7375  Shallow 5.4 magnitude earthquake rattles centr...   
1010   9097                   ICE Agent Commits Suicide in NYC   
1043   9203        Political Correctness for Yuengling Brewery   
1218   1602  Poll gives Biden edge over Clinton against GOP...   
1438   4562                  Russia begins airstrikes in Syria   
1493   4748                                         Trump &amp   
1591   3508                         Belgian police mount raids   
1630   7559  STATE OF GEORGIA FIRES PASTOR BECAUSE OF HIS F...   
1716   3634      The Latest On Paris Attack: Manhunt Continues   
1900   8470  The Amish In America Commit Their Vote To Dona...   
1968   6404  #BREAKING: SECOND Assassination Attempt On Tru...   
2176  10499  30th Infantry Division: “Work Horse of the Wes...   
2184      9               Planned Parenthood’s lobbying effort   
2493  10492  TOP BRITISH GENERAL WARNS OF NUCLEAR WAR WITH ...   
2549  10138                Inside The Mind Of An FBI Informant   
2880   4953       Gary Johnson Avoids Typical Third-Party Fade   
2920    496                   Nearly 300K New Jobs In February   
3010   5741                                      Why Trump Won   
3069   4131    Jesse Matthew charged in Hannah Graham's murder   
3110   8748      WATCH: Mass Shooting Occurs During #TrumpRiot   
3130   6717                   Jim Rogers: It’s Time To Prepare   
3210   2943              Islamic State admits defeat in Kobani   
3372   5248        Clinton Cries Racism Tagging Trump with KKK   
3478   3624           Suspects In Paris Magazine Attack Killed   
3537   6268  Chart Of The Day: Since 2009—–Recovery For The 5%   
3578   2738                              Ted Cruz launches bid   
3649   4025  State Dept. IDs 2 Americans killed in Nepal quake   
3706   9954  Incredible smoke haze seen outside NDTV office...   

                                                   text  \
192                         Democrats Lose In The South   
308                    Leonardo DiCaprio to the rescue?   
382                                 human remains found   
660                                  U.S. investigating   
889                            US issues travel warning   
911                            shakes buildings in Rome   
1010   Leaves Note Revealing Gov’t Plans to Round-up...   
1043                    What About Our Opioid Epidemic?   
1218                               VP meets with Trumka   
1438             U.S. warns of new concerns in conflict   
1493   Clinton Were Very Convincing...on How Lousy t...   
1591       prosecutors acknowledge missed opportunities   
1630       GOVERNMENT DIDN’T “APPROVE” BIBLICAL SERMONS   
1716                       Brothers Were On No-Fly List   
1900   Mathematically Guaranteeing Him A Presidentia...   
1968                       Suspect Detained (LIVE BLOG)   
2176                             The Big Picture TV-211   
2184                     pay raises for federal workers   
2493                    “THE END OF LIFE AS WE KNOW IT”   
2549          Terri Linnell Admits Role As Gov’t Snitch   
2880                    Best Polling Since Perot in ‘92   
2920                   Unemployment Dips To 5.5 Percent   
3010                                   Why Clinton Lost   
3069                   DA will not pursue death penalty   
3110                              Media Ignores (Video)   
3130   Economic And Financial Collapse Imminent (VIDEO)   
3210                                  blames airstrikes   
3372                              Trump Says 'She Lies'   
3478             Market Gunman And 4 Hostages Also De

### 2.2 Deleting the X1 and X2
Since X1 and X2 have 33 occurences in total and there is no clear structure to fit them within existing data we decided to delete them

In [20]:
# delete X1 and X2 variables
df_train = df_train[df_train.X1.isnull()] 

In [21]:
# ALTERNATIVE COLUMN DELETE 
df_train = df_train.drop("X1",axis=1)
df_train = df_train.drop("X2",axis=1)

In [22]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3966 entries, 0 to 3998
Data columns (total 4 columns):
ID       3966 non-null int64
title    3966 non-null object
text     3966 non-null object
label    3966 non-null object
dtypes: int64(1), object(3)
memory usage: 154.9+ KB


## 3. Prepare Dataset

### 3.1 Cross Validation

In [23]:
# create training and testing vars
X_train, X_test, y_train, y_test = train_test_split(df_train.text, df_train.label)
print (X_train.shape, y_train.shape)
print (X_test.shape, y_test.shape)

(2974,) (2974,)
(992,) (992,)


### 3.2 Feature Extraction
Bag of Words

In [24]:
# Feature Extraction
from sklearn.feature_extraction.text import CountVectorizer

tf_weighting = CountVectorizer()
X_train_counts = tf_weighting.fit_transform(df_train.text)
X_train_counts.shape

(3966, 55312)

### 3.2 TF-IDF
Assigning weights to words in documents by dividing the count of words by the lenght of the document and reducing the weight of most common words

In [25]:
#TF-IDF
from sklearn.feature_extraction.text import TfidfTransformer

tfidf_transformer = TfidfTransformer()
X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)
X_train_tfidf.shape

(3966, 55312)

## 4. SVM 

#### First attempt - SGDClassifier

In [26]:
from sklearn.linear_model import SGDClassifier
text_clf_svm = Pipeline([('vect', CountVectorizer()), ('tfidf', TfidfTransformer()),
                         ('clf-svm', SGDClassifier(loss='squared_loss', penalty='l2',alpha=1e-3, n_iter=5, random_state=42))])

text_clf_svm = text_clf_svm.fit(X_train, y_train)
predicted_svm = text_clf_svm.predict(X_test)
np.mean(predicted_svm == y_test)

0.9153225806451613

#### SVM including removing the stop words  // Lower the result // 

In [27]:
# Training Support Vector Machines - SVM and calculating its performance
from sklearn.linear_model import SGDClassifier
text_clf_svm = Pipeline([('vect', CountVectorizer(stop_words='english')), 
                         ('tfidf', TfidfTransformer()),
                         ('clf-svm', SGDClassifier(loss='hinge', penalty='l2',alpha=1e-3, n_iter=5, random_state=42))])

text_clf_svm = text_clf_svm.fit(X_train, y_train)
predicted_svm = text_clf_svm.predict(X_test)
np.mean(predicted_svm == y_test)

0.9112903225806451

### 4.1  Applying LinearSCV - score improvement 

In [28]:
from sklearn.linear_model import SGDClassifier
from sklearn.svm import LinearSVC

text_clf_svm2 = Pipeline([('vect', CountVectorizer()), ('tfidf', TfidfTransformer()),
                         ('clf-svm', LinearSVC())])

text_clf_svm2 = text_clf_svm2.fit(X_train, y_train)
predicted_svm2 = text_clf_svm2.predict(X_test)
np.mean(predicted_svm2 == y_test)

0.9243951612903226

### 4.2 Grid Search for SVM

In [ ]:

from sklearn.model_selection import GridSearchCV
parameters_svm = {'vect__ngram_range': [(1, 1), (1, 2)], 'tfidf__use_idf': (True, False),'clf-svm__alpha': (1e-2, 1e-3)}

gs_clf_svm = GridSearchCV(text_clf_svm, parameters_svm, n_jobs=-1)
gs_clf_svm = gs_clf_svm.fit(X_train, y_train)


print(gs_clf_svm.best_score_)
print(gs_clf_svm.best_params_)

# 5. Naive Bayes

In [19]:
# Training Naive Bayes (NB) classifier on training data.
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB().fit(X_train_tfidf, df_train.label)

In [20]:
# Pipeline - standard
from sklearn.pipeline import Pipeline

text_clf = Pipeline([('vect', CountVectorizer()), ('tfidf', TfidfTransformer()), ('clf', MultinomialNB())])

text_clf = text_clf.fit(X_train, y_train)

In [21]:
# Performance - standard
predicted = text_clf.predict(X_test)
np.mean(predicted == y_test)

0.7368951612903226

In [22]:
# Pipeline with adjusting MultinomialNB

text_clf2 = Pipeline([('vect', CountVectorizer()), ('tfidf', TfidfTransformer()), ('clf', MultinomialNB(fit_prior=False))])

text_clf2 = text_clf2.fit(X_train, y_train)

In [23]:
# Performance with adjusting MultinomialNB
predicted = text_clf2.predict(X_test)
np.mean(predicted == y_test)

0.7520161290322581

### 4.1 Grid Search for Naive Bayes
I tried to optimize the parameters for Grid Search (estimator and scoring) : 
http://scikit-learn.org/stable/modules/grid_search.html#grid-search

In [24]:
# Original Search
from sklearn.model_selection import GridSearchCV
parameters = {'vect__ngram_range': [(1, 1), (1, 2)], 'tfidf__use_idf': (True, False), 'clf__alpha': (1e-2, 1e-3)}

gs_clf = GridSearchCV(text_clf, parameters, n_jobs=-1)
gs_clf = gs_clf.fit(X_train, y_train)

print(gs_clf.best_score_)
print(gs_clf.best_params_)

0.8971082716879624
{'clf__alpha': 0.001, 'tfidf__use_idf': False, 'vect__ngram_range': (1, 1)}


In [25]:
# Search 2 - Changing parameters based on the previous results. Slight Improvement
from sklearn.model_selection import GridSearchCV
parameters = {'vect__ngram_range': [(1, 1), (1, 5)], 'tfidf__use_idf': (True, False), 'clf__alpha': (0.0001, 0.001, 0.01)}

gs_clf = GridSearchCV(text_clf, parameters, n_jobs=-1)
gs_clf = gs_clf.fit(X_train, y_train)

print(gs_clf.best_score_)
print(gs_clf.best_params_)

0.8971082716879624
{'clf__alpha': 0.001, 'tfidf__use_idf': False, 'vect__ngram_range': (1, 1)}


In [26]:
# Search 3
from sklearn.model_selection import GridSearchCV
parameters = {'vect__ngram_range': [(1, 4), (1, 5), (1, 6)], 'tfidf__use_idf': (True, False), 'clf__alpha': (0.001, 0.005, 0.01)}

gs_clf = GridSearchCV(text_clf, parameters, n_jobs=-1)
gs_clf = gs_clf.fit(X_train, y_train)

print(gs_clf.best_score_)
print(gs_clf.best_params_)

0.8967720242098184
{'clf__alpha': 0.001, 'tfidf__use_idf': True, 'vect__ngram_range': (1, 6)}


### 4.2 Stemming for Naive Bayes

In [27]:
from nltk.stem.snowball import SnowballStemmer
stemmer = SnowballStemmer("english", ignore_stopwords=True)

class StemmedCountVectorizer(CountVectorizer):
    def build_analyzer(self):
        analyzer = super(StemmedCountVectorizer, self).build_analyzer()
        return lambda doc: ([stemmer.stem(w) for w in analyzer(doc)])
    
stemmed_count_vect = StemmedCountVectorizer(stop_words='english')

text_mnb_stemmed = Pipeline([('vect', stemmed_count_vect), 
                             ('tfidf', TfidfTransformer()), 
                             ('mnb', MultinomialNB(fit_prior=False))])

text_mnb_stemmed = text_mnb_stemmed.fit(X_train, y_train)

predicted_mnb_stemmed = text_mnb_stemmed.predict(X_test)

np.mean(predicted_mnb_stemmed == y_test)

0.8064516129032258

# 6. Logistic Regression

#### 6.1 Repeating the dataset preparation for Logistic Regression
We found out that the transformation process for logistic regression is different than for the other two models. Applying same feature extraction is causing a crash for one of the parts and for now the only way around it we found is repeating and adjusting the process.

In [28]:
# Cross Validation
X_train, X_test, y_train, y_test = train_test_split(df_train.text, df_train.label)
print (X_train.shape, y_train.shape)
print (X_test.shape, y_test.shape)

# Feature Extraction - Logistic Regression
from sklearn.feature_extraction.text import CountVectorizer
tf_weighting = CountVectorizer()
X_train_counts = tf_weighting.fit_transform(X_train)
X_test_counts = tf_weighting.transform(X_test)

#TF-IDF - Logistic Regression
from sklearn.feature_extraction.text import TfidfTransformer
tfidf_transformer = TfidfTransformer()
X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)
X_train_tfidf.shape

X_test_tfidf = tfidf_transformer.transform(X_test_counts)

(2974,) (2974,)
(992,) (992,)


### 6.2 Logistic Regression Model

In [29]:
lr_count_clf = LogisticRegression()

In [30]:
lr_count_clf.fit(X_train_counts, y_train)
prediction = lr_count_clf.predict(X_test_counts)
result0 = metrics.accuracy_score(y_test, prediction)
print("logistic regression accuracy: %0.3f" % result0)

result1 = cross_val_score(lr_count_clf, X_train_counts, y_train, scoring="accuracy", cv=5).mean()
print("logistic regression cross-validation accuracy: %0.3f" % result1)

logistic regression accuracy: 0.923
logistic regression cross-validation accuracy: 0.899


In [31]:
X_train, X_test, y_train, y_test = train_test_split(df_train.text, df_train.label)

tf_weighting = CountVectorizer()
X_train_counts = tf_weighting.fit_transform(df_train.text)
X_train_counts.shape

tfidf_transformer = TfidfTransformer()
X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)
X_train_tfidf.shape

(3966, 55312)

In [32]:
from sklearn.pipeline import Pipeline

text_clf_logreg = Pipeline([('vect', CountVectorizer()), ('tfidf', TfidfTransformer()), ('clf2', LogisticRegression())])

text_clf_logreg = text_clf_logreg.fit(X_train, y_train)


# Performance - standard
predicted = text_clf_logreg.predict(X_test)
np.mean(predicted == y_test)

0.8931451612903226

## 7. PassiveAgressive Classifier
https://github.com/docketrun/Detecting-Fake-News-with-Scikit-Learn/blob/master/Attempting%20to%20detect%20fake%20news.ipynb

In [33]:
# Preparation
count_vectorizer = CountVectorizer(stop_words='english')
count_train = count_vectorizer.fit_transform(X_train)
count_test = count_vectorizer.transform(X_test)

tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_df=0.7)
tfidf_train = tfidf_vectorizer.fit_transform(X_train)
tfidf_test = tfidf_vectorizer.transform(X_test)

In [34]:
# Importing and adjusting the PAC
from sklearn.linear_model import PassiveAggressiveClassifier
linear_clf = PassiveAggressiveClassifier(n_iter=100, fit_intercept=False, warm_start=True)

In [35]:
linear_clf.fit(tfidf_train, y_train) #
pred = linear_clf.predict(tfidf_test)
score = metrics.accuracy_score(y_test, pred)
print("accuracy:   %0.3f" % score)
cm = metrics.confusion_matrix(y_test, pred, labels=['FAKE', 'REAL'])

accuracy:   0.917


## 8. SUMMARY

#### We have tried several models - found at different sources. For our specific approach the SVM with LinearSVC gave the best score around 0.924

#### Resources:
https://towardsdatascience.com/machine-learning-nlp-text-classification-using-scikit-learn-python-and-nltk-c52b92a7c73a
https://github.com/docketrun/Detecting-Fake-News-with-Scikit-Learn/blob/master/Attempting%20to%20detect%20fake%20news.ipynb
https://bbengfort.github.io/tutorials/2016/05/19/text-classification-nltk-sckit-learn.html
http://scikit-learn.org/stable/auto_examples/model_selection/plot_multi_metric_evaluation.html#sphx-glr-auto-examples-model-selection-plot-multi-metric-evaluation-py
